### ISSUE TITLE
Nested json is transformed to string when used “$type” key in object

### ISSUE THREAD

**Issue by tomasz-py** [prisma/prisma#21454]
### Bug description

Json value which has array of objects where one of keys is "$type" is transformed to string

`
  const data = await prisma.test.create({
    data: {
       value: {
         options: [{ $type: "z" }, { $type: 1 }],
        },
    },
  });
`

data.value will looks like: 
`{ options: [ '{"$type":"z"}', '{"$type":1}' ] }`

instead of 
`{ options: [ {'$type': 'z'}, {'$type': 1 } ] }`

the same situation is for object in object:
`  const data = await prisma.test.create({
    data: {
         value: {
          options: {
            $type: "test",
          },
        },
    },
  });`

data.value will contain options as string instead of json
`{ options: '{"$type":"test"}' }`

Additional info:
its stored properly if object key is $test, so its only breaks when key is "$type"

### How to reproduce

1. Use the following schema
2. Generate the client and run the following script
3. Check value in database

### Expected behavior

Prisma should store array of jsons as jsons instead of transforming them into string

### Prisma information

```prisma
model Test {
  id  Int  @id @default(autoincrement())
  value  Json?  @db.JsonB
}
```

```ts
import { PrismaClient } from "@prisma/client";

async function main() {
  const prisma = new PrismaClient();

  const data = await prisma.test.create({
    data: {
       value: {
         options: [{ $type: "z" }, { $type: 1 }],
        },
    },
  });

 const otherData =  await prisma.test.create({
    data: {
        value: {
          options: {
            $type: "test",
          },
        },
    },
  });

  console.log(data.value); // { options: [ '{"$type":"z"}', '{"$type":1}' ] }
 
  console.log(otherData.value) // { options: '{"$type":"test"}' }

  prisma.$disconnect();
}

main();
```

### Environment & setup

- OS: Windows 11
- Database: PostgreSQL
- Node.js version: v20.5.1

### Prisma Version

```
prisma                  : 5.1.1
@prisma/client          : 5.1.1
Current platform        : windows
Query Engine (Node-API) : libquery-engine 6a3747c37ff169c90047725a05a6ef02e32ac97e (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Schema Engine           : schema-engine-cli 6a3747c37ff169c90047725a05a6ef02e32ac97e (at node_modules\@prisma\engines\schema-engine-windows.exe)
Schema Wasm             : @prisma/prisma-schema-wasm 5.1.1-1.6a3747c37ff169c90047725a05a6ef02e32ac97e
Default Engines Hash    : 6a3747c37ff169c90047725a05a6ef02e32ac97e
Studio                  : 0.492.0
```

**Comment by JosephHalter** [comment]
Thank you, I thought that I was getting crazy but indeed it's easy to reproduce, just try saving an array of objects with $type key in a json column and you'll see that the objects are serialized as strings instead, making it impossible to query inside or even read back without having to use JSON.parse recursively.

**Comment by tomasz-py** [comment]
Thanks for confirmation. 
Its critical bug for us because we have migrated to Prisma now and we found out that it will destroy our old data in database.

**Comment by niba** [comment]
We are having the same issue. We migrated to Prisma and we can't fetch data from one of our tables because when you fetch data with a structure containing "$type" then it explodes with error: `Unknown tagged value`.
I see that it is probably because of jsonProtocol implementation.

Do you have any workaround for this bug? @janpio 

This issue is pretty critical for us, as there's no easy fix on our side.

**Comment by karanssj4** [comment]
found a workaround.
set toJSON on the prototype of the object you want to insert.
This shouldn't have any problems for puse JS objects. maybe someone from the prisma team should confirm this? @janpio 

change from

```
 const valueToInsert =  {
   value: {
      options: [{ $type: "z" }, { $type: 1 }],
    },
  }
};

Object.setPrototypeOf(valueToInsert, {
  toJSON() { return valueToInsert }
});

const data = await prisma.test.create({
  data: valueToInsert
});
  ```

**Comment by niba** [comment]
@karanssj4 
nice, I tried to use toJSON but it didn't work in my case. I need to check again

another problem is that you can't fetch this kind of data from database because you get exception error :<
try to use findMany etc 

$types is used in their implementation of JSONProtocol somewhere in rust engine and it goes crazy when you have key like that

I don't understand how this bug is not high priority right now

**Comment by karanssj4** [comment]
@niba i've not seen this issue while reading back from db, only while inserting it. we used raw queries for insrting and prisma crud methods for reading, it worked

can you share the read opearation that messes it up for you?

**Comment by SevInf** [comment]
Reproduced the issue in #22714

**Comment by SevInf** [comment]
Hello @tomasz-py @karanssj4 @niba.
The issue is fixed and the fix will be released in a version `5.9.0` tomorrow. 
If you want to test dev snapshot, you can try `5.9.0-dev.64` version. We don't recommend you use dev snapshots in production, but it is enough to verify that fix works for your case.
